sstables/sstable_set.cc

```diff
@@ -61,94 +61,422 @@ std::ostream& operator<<(std::ostream& os, const sstables::sstable_run& run) {
+
+
+std::vector<shared_sstable> sstable_set::select(const dht::partition_range& range) const {
+    return _all->version().select(range);
+}
+
+// Return all runs which contain any of the input sstables.
+std::vector<sstable_run> sstable_set::select_sstable_runs(const std::vector<shared_sstable>& sstables) const {
+    return _all->version().select_sstable_runs(sstables);
+}
+
+lw_shared_ptr<const sstable_list> sstable_set::all() const {
+    return _all;
+}
+
+void sstable_set::insert(shared_sstable sst) {
+    _all->insert(sst);
+}
+
+void sstable_set::erase(shared_sstable sst) {
+    _all->erase(sst);
 }
 
 sstable_set::incremental_selector::~incremental_selector() = default;
 
 sstable_set::incremental_selector::incremental_selector(sstable_set::incremental_selector&&) noexcept = default;
 
-sstable_set::incremental_selector::selection
-sstable_set::incremental_selector::select(const dht::ring_position_view& pos) const {
+sstable_set::incremental_selector::incremental_selector(std::unique_ptr<incremental_selector_impl> impl, const schema& s, lw_shared_ptr<sstable_list> sstl)
+    : _impl(std::move(impl))
+    , _cmp(s)
+    , _sstl(std::move(sstl)) {
+}
+
+sstable_set::incremental_selector::selection sstable_set::incremental_selector::select(const dht::ring_position_view& pos) const {
     if (!_current_range_view || !_current_range_view->contains(pos, _cmp)) {
-        std::tie(_current_range, _current_sstables, _current_next_position) = _impl->select(pos);
+        std::vector<shared_sstable> _current_versioned_sstables;
+        std::tie(_current_range, _current_versioned_sstables, _current_next_position) = _impl->select(pos);
+        _current_sstables = boost::copy_range<std::vector<shared_sstable>>(_current_versioned_sstables
+                             | boost::adaptors::filtered([this] (shared_sstable sst) { return _sstl->contains(sst); })
+                             | boost::adaptors::transformed([] (shared_sstable sst) { return sst; }));
         _current_range_view = _current_range->transform([] (const dht::ring_position& rp) { return dht::ring_position_view(rp); });
     }
     return {_current_sstables, _current_next_position};
 }
 
-sstable_set::incremental_selector
-sstable_set::make_incremental_selector() const {
-    return incremental_selector(_impl->make_incremental_selector(), *_schema);
+sstables::sstable_set::incremental_selector sstable_set::make_incremental_selector() const {
+    return incremental_selector(_all->version().make_incremental_selector(), *_all->version().get_schema(), _all);
+}
+
+sstable_set_version::~sstable_set_version() {
+    for (auto&& added_sst : _added) {
+        _versioned_data->remove(added_sst);
+    }
+    if (_prev) {
+        _prev->_next.erase(this);
+    }
+}
+
+sstable_set_version::sstable_set_version(std::unique_ptr<sstable_set_impl> impl, schema_ptr schema) 
+    : _versioned_data(make_lw_shared<versioned_sstable_set_data>(std::move(impl))) 
+    , _schema(std::move(schema)) {
```

Comment:
Shouldn't we populate _times_added too? Why is this constructor different than calling insert() on every sstable?

Reply:
We're assuming that an sstable_set is constructed from an empty sstable_set_impl. We were assuming that before the whole patchset too but it wasn't specified anywhere - I've added checks for it in the second patch of the series.
Considering I'm modifying the entire sstable_set I could probably change this constructor to work on any sstable_set_impl, but I'm not sure if that may ever be needed

Reply:
Just add a comment.

Reply:
This also refers to an older commit, but in the newest version there is no indication that the set must be empty in this constructor(sstable_set_version). Instead, I've added a logic_error that checks it in the sstable_set constructor. Do you think a comment would be useful here too?

Reply:
I'm lost in the details, but a comment can't hurt [1], whereas the lack of a comment can.

[1] a wrong comment can hurt

Reply:
Ok, I'm adding comments in both places then